### PR TITLE
[FLINK-16806][operators] Add support for InputSelection to MultipleInputStreamOperator

### PR DESCRIPTION
**Pull request by pnowojski**
This PR adds support for InputSelection to MultipleInputStreamOperator.

## Brief change log

Please check individual commit messages.

## Verifying this change

This change is covered by existing tests like `MultipleInputITCase` and adds couple of new tests to cover the new feature.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java [hunk 1]
```diff
@@ -112,7 +112,7 @@ int selectNextInputIndex(int lastReadInputIndex) {
 	}
 
 	boolean shouldSetAvailableForAnotherInput() {
-		return availableInputsMask != allSelectedMask && inputSelection.areAllInputsSelected();
+		return (inputSelection.getInputMask() & allSelectedMask & ~availableInputsMask) != 0;
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: Could we avoid 1 `& allSelectedMask` here? It feels like it should only be relevant for `ALL`. We could then eagerly apply `allSelectedMask` instead of setting ALL (`new InputSelection(-1 & allSelectedMask)`).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java [hunk 2]
```diff
@@ -112,7 +112,7 @@ int selectNextInputIndex(int lastReadInputIndex) {
 	}
 
 	boolean shouldSetAvailableForAnotherInput() {
```
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: meta note: it's really confusing to have inputSelection, InputSelectable, and inputSelector. Could you try to consolidate that?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		selectionHandler.nextSelection();
+
+		assertFalse(selectionHandler.shouldSetAvailableForAnotherInput());
+
+		selectionHandler.setUnavailableInput(0);
+		assertFalse(selectionHandler.shouldSetAvailableForAnotherInput());
+
+		selectionHandler.setUnavailableInput(2);
+		assertTrue(selectionHandler.shouldSetAvailableForAnotherInput());
+
```
Comment (AHeise) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandlerTest.java]: Add a check what happens on `selectionHandler.setAvailableInput(2)`?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMailboxTestHarness.java [hunk 4]
```diff
@@ -122,6 +122,12 @@ private void maybeProcess() throws Exception {
 
 	public void process() throws Exception {
 		while (streamTask.inputProcessor.isAvailable() && streamTask.mailboxProcessor.isMailboxLoopRunning()) {
+			processSingleStep();
```
Comment (AHeise) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMailboxTestHarness.java]: Kind of replicates the checks. Would it make sense to have a `boolean processSingleStep`, where we can just execute a `while(processSingleStep()) ;`?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+			testBase(new TestReadFinishedInputStreamOperatorFactory(), true, new ArrayDeque<>(), true);
+			fail("should throw an IOException");
+		} catch (Exception t) {
+			if (!ExceptionUtils.findThrowableWithMessage(t, "Can not make a progress: all selected inputs are already finished").isPresent()) {
+				throw t;
+			}
+		}
+	}
+
+	private void testBase(
```
Comment (AHeise) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java]: Not a big fan of the name. How about `testInputSelection`?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * Test selective reading.
+ */
+public class StreamTaskMultipleInputSelectiveReadingTest {
+
+	@Test
+	public void testAnyOrderedReading() throws Exception {
```
Comment (AHeise) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java]: Tests are pretty hard to follow as input is separated from output. I don't have a good solution other than replicating that in each test. You could also extract the input sequence as a constant (array). In this way, it's immediately visible when looking at the test (from top to bottom).

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+		ArrayDeque<Object> expectedOutput = new ArrayDeque<>();
+		expectedOutput.add(new StreamRecord<>("[1]: Hello-1"));
+		expectedOutput.add(new StreamRecord<>("[2]: 1"));
+		expectedOutput.add(new StreamRecord<>("[1]: Hello-2"));
+		expectedOutput.add(new StreamRecord<>("[2]: 2"));
+		expectedOutput.add(new StreamRecord<>("[1]: Hello-3"));
+		expectedOutput.add(new StreamRecord<>("[2]: 3"));
+		expectedOutput.add(new StreamRecord<>("[2]: 4"));
+
+		testBase(new TestAnyModeMultipleInputStreamOperator.Factory(), false, expectedOutput, true);
```
Comment (AHeise) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java]: It's again not easy to see what all the operators are doing. Would it be possible to have a generic operator, where you pass a `InputSelectable`, such that it is defined at the same place as the expected output?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java (hunk 2 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: Do you have some proposal? `InputSelectable` is an interface for users to implement, to provide the chosen `InputSelection`. That gives as `inputSelector` from our perspective. Now we are wrapping some input selection handling logic into classes like `MultipleInputSelectionHandler`. How would you suggest to change it?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java (hunk 1 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: What's the problem? I think performance wise it doesn't matter, the `& allSelectedMask` would still need to be there somewhere on the critical path. 

From logic perspective, I actually prefer keeping `ALL` as valid value for the `this.inputSelection` and doing the `& allSelectedMask` here, as those extra bits are not an issue in general, but only for this particular line of code (`!= 0` comparison).

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandlerTest.java (hunk 3 above)
Comment (pnowojski) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandlerTest.java]: It's available by default, so the previous check covering that.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java (hunk 7 above)
Comment (pnowojski) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java]: Hmmm, I think that would be quite difficult to abstract. I get your point, but I'm not sure if it's worth it. (especially take a look at the logic inside `SpecialRuleReadingStreamOperatorFactory`).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: > What's the problem? I think performance wise it doesn't matter, the `& allSelectedMask` would still need to be there somewhere on the critical path.

The point is that it would be moved out of the critical path. You only need to do it once in the ctor.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java (hunk 2 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: Can't we keep the name of `InputSelectable` `inputSelectable`? It's especially tricky to distinguish `inputSelector` and `inputSelection` (edit distance 2).

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandlerTest.java (hunk 3 above)
Comment (AHeise) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandlerTest.java]: I was more suggesting to test the state change to go back to default.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java (hunk 7 above)
Comment (AHeise) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java]: Yes the logic would be hard to generalize. It may be worth just to do it partially for the easy cases.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java (hunk 1 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: Unfortunately no. We would still need to do this action once per record around here:
```
org.apache.flink.streaming.runtime.io.MultipleInputSelectionHandler#nextSelection
	void nextSelection() {
		if (inputSelector == null) {
			inputSelection = InputSelection.ALL;
		} else {
			inputSelection = inputSelector.nextSelection();
		}
	}
```
? And I would doubt that saving one bit operation would be visible anywhere, while as I wrote logically I think it's clearer to have this operation here.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java (hunk 2 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: `inputSelectable ` is consistent but It's a bit worse description what it does. But ok, I can change it if you prefer so. (will be a commit on top)

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java (hunk 7 above)
Comment (pnowojski) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java]: But `TestAnyModeMultipleInputStreamOperator` is the only easy case that can be easily generalised. Unifying it with `TestSequentialMultipleInputStreamOperator` would basically require to pass the whole operator here, via the generalised factory. I really don't think it makes sense And generalising just `TestAnyModeMultipleInputStreamOperator` also doesn't make sense, as that would just add code complexity?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: Okay then keep as is.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java (hunk 2 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: then please have the extra commit and I will check if readability improves for me. I also got your point, but I was really confused initially.

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/MultipleInputSelectionHandler.java]: Ops sorry, I deleted my comment and replaced it with just a thumb up, as I changed my mind. I like the `inputSelectable` name for this field better :)

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java (hunk 7 above)
Comment (AHeise) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskMultipleInputSelectiveReadingTest.java]: Okay keep as is.

**Comment by pnowojski** [comment]
Thanks for the review @AHeise . Merging. Azure test failure is unrelated: https://issues.apache.org/jira/browse/FLINK-16821
